apps/builder/app/builder/features/props-panel/use-props-logic.ts

```diff
@@ -1,156 +1,240 @@
 import { useState } from "react";
 import { nanoid } from "nanoid";
 import type { Instance, Prop } from "@webstudio-is/project-build";
-import type { WsComponentPropsMeta } from "@webstudio-is/react-sdk";
 import { getComponentPropsMeta } from "@webstudio-is/react-sdk";
+import type { PropMeta, PropValue } from "./shared";
+
+type PropOrName = { prop?: Prop; propName: string };
+export type PropAndMeta = { prop?: Prop; propName: string; meta: PropMeta };
+export type NameAndLabel = { name: string; label?: string };
+
+// The value we set prop to when it's added
+export const getStartingValue = (meta: PropMeta): PropValue | undefined => {
+  if (meta.type === "string" && meta.control !== "file-image") {
+    return {
+      type: "string",
+      value: meta.defaultValue ?? "",
+    };
+  }
 
-export type UserPropValue = Prop extends infer T
-  ? T extends { value: unknown; type: unknown }
-    ? { value: T["value"]; type: T["type"] }
-    : never
-  : never;
-
-export const getValueFromPropMeta = (
-  propValue?: WsComponentPropsMeta["props"][string]
-) => {
-  let typedValue: UserPropValue = {
-    type: "string",
-    value: `${propValue?.defaultValue ?? ""}`,
-  };
+  if (meta.type === "number") {
+    return {
+      type: "number",
+      value: meta.defaultValue ?? 0,
+    };
+  }
 
-  if (propValue?.control === "boolean") {
-    typedValue = {
+  if (meta.type === "boolean") {
+    return {
       type: "boolean",
-      value: propValue.defaultValue ?? false,
+      value: meta.defaultValue ?? false,
     };
   }
 
-  if (propValue?.control === "number") {
-    typedValue = {
-      type: "number",
-      value: propValue.defaultValue ?? 0,
+  if (meta.type === "string[]") {
+    return {
+      type: "string[]",
+      value: meta.defaultValue ?? [],
     };
   }
 
-  return typedValue;
+  // If undefined is returned,
+  // we will not add a prop in storage until we get an onChange from control.
+  //
+  // User may have this experience:
+  //   - they added a prop but didn't touch the control
+  //   - they closed props panel
+  //   - when they open props panel again, the prop is not there
+  //
+  // We want to avoid this if possible, but for some types like "asset" we can't
+  return undefined;
+};
+
+const getStartingProp = (
+  instanceId: Instance["id"],
+  meta: PropMeta,
+  name: string
+): Prop | undefined => {
+  const value = getStartingValue(meta);
+  return value && { id: nanoid(), instanceId, name, ...value };
+};
+
+const getDefaultMetaForType = (type: Prop["type"]): PropMeta => {
```

Comment:
seems like this could be just a map from type to meta instead of a function with all the ifs

e.g. 
```
typeToMetaMap = {asset: { type: "string", control: "file-image", required: false }}
```
Also, where is the type "asset" coming from, storybook calls it "file", wouldn't it make sense to call it file too? or is it deeply depending on our internal asset?

Reply:
wrapping into a function is fine, but ifs instead of a simple map looks fishy, because you don't seem to need custom logic there, its just a switch/case type of thing going on here

Reply:
re-wrote with switch/case